tests/query/snql/test_joins.py

```diff
@@ -0,0 +1,237 @@
+import pytest
+import uuid
+from typing import Sequence, Tuple, Union
+
+from snuba.datasets.entities import EntityKey
+from snuba.datasets.entities.factory import get_entity
+from snuba.query.data_source.join import (
+    IndividualNode,
+    JoinClause,
+    JoinClass,
+    JoinCondition,
+    JoinConditionExpression,
+    JoinModifier,
+    JoinRelationship,
+    JoinType,
+)
+from snuba.query.data_source.simple import Entity as QueryEntity
+from snuba.query.snql.joins import build_join_clause, EntityTuple, RelationshipTuple
+
+###################################
+## Used to build expected result ##
+###################################
+
+
+def node(alias: str, name: str) -> IndividualNode[QueryEntity]:
+    return IndividualNode(
+        alias,
+        QueryEntity(
+            EntityKey(name), get_entity(EntityKey(name)).get_data_model(), alias
+        ),
+    )
+
+
+def join_clause(
+    lhs_alias: str, lhs: Union[str, JoinClause[QueryEntity]], rhs: str
+) -> JoinClause[QueryEntity]:
+    rhs_alias, rhs = rhs.split(":", 1)
+    return JoinClause(
+        left_node=node(lhs_alias, lhs) if isinstance(lhs, str) else lhs,
+        right_node=node(rhs_alias, rhs),
+        keys=[
+            JoinCondition(
+                left=JoinConditionExpression(lhs_alias, "event_id"),
+                right=JoinConditionExpression(rhs_alias, "event_id"),
+            )
+        ],
+        join_type=JoinType.INNER,
+        join_modifier=JoinModifier.SEMI,
+    )
+
+
+test_cases = [
+    pytest.param(
+        [("ev:events", "t:transactions")],
+        join_clause("ev", "events", "t:transactions"),
+        id="simple",
+    ),
+    pytest.param(
+        [("ev:events", "t:transactions"), ("t:transactions", "er:errors")],
+        join_clause("t", join_clause("ev", "events", "t:transactions"), "er:errors"),
+        id="depth=2",
+    ),
+    pytest.param(
+        [("t:transactions", "er:errors"), ("ev:events", "t:transactions")],
+        join_clause("t", join_clause("ev", "events", "t:transactions"), "er:errors"),
+        id="depth=2 bottom-up",
+    ),
```

Comment:
Thanks. It is important that no matter the order the cliient provides the relationship into, the join structure will be the same.